File: src/accessiweather/shortcut_preferences.py
"""Helpers for configurable window and tray shortcuts."""

from __future__ import annotations

from dataclasses import dataclass

DEFAULT_SHOW_MAIN_WINDOW_SHORTCUT = "Ctrl+Shift+W"
DEFAULT_HIDE_MAIN_WINDOW_SHORTCUT = "Ctrl+Shift+M"
DEFAULT_READ_TRAY_INFO_SHORTCUT = "Ctrl+Shift+I"

_MODIFIER_ALIASES = {
    "ALT": "Alt",
    "CTRL": "Ctrl",
    "CONTROL": "Ctrl",
    "CMD": "Ctrl",
    "COMMAND": "Ctrl",
    "OPTION": "Alt",
    "SHIFT": "Shift",
}
_SPECIAL_KEYS = {
    "ESC": "Escape",
    "ESCAPE": "Escape",
    "SPACE": "Space",
    "TAB": "Tab",
}
_VISIBLE_MODIFIER_ORDER = ("Ctrl", "Alt", "Shift")
_ACCELERATOR_MODIFIER_NAMES = {
    "Ctrl": "ACCEL_CTRL",
    "Alt": "ACCEL_ALT",
    "Shift": "ACCEL_SHIFT",
}
_HOTKEY_MODIFIER_NAMES = {
    "Ctrl": ("MOD_CONTROL", "ACCEL_CTRL"),
    "Alt": ("MOD_ALT", "ACCEL_ALT"),
    "Shift": ("MOD_SHIFT", "ACCEL_SHIFT"),
}
_SPECIAL_KEY_CODE_NAMES = {
    "Escape": ("WXK_ESCAPE", 27),
    "Space": ("WXK_SPACE", 32),
    "Tab": ("WXK_TAB", 9),
}


@dataclass(frozen=True)
class ShortcutPreference:
    setting_name: str
    label: str
    default: str
    handler_name: str


@dataclass(frozen=True)
class ShortcutBinding:
    modifiers: tuple[str, ...]
    key: str

    @property
    def normalized(self) -> str:
        parts = [*self.modifiers, self.key]
        return "+".join(parts)

    def accelerator_flags(self, wx_module) -> int:
        flags = getattr(wx_module, "ACCEL_NORMAL", 0)
        for modifier in self.modifiers:
            flags |= getattr(wx_module, _ACCELERATOR_MODIFIER_NAMES[modifier], 0)
        return flags

    def hotkey_modifiers(self, wx_module) -> int:
        flags = 0
        for modifier in self.modifiers:
            primary_name, fallback_name = _HOTKEY_MODIFIER_NAMES[modifier]
            flags |= getattr(wx_module, primary_name, getattr(wx_module, fallback_name, 0))
        return flags

    def key_code(self, wx_module) -> int:
        if len(self.key) == 1:
            return ord(self.key)
        if self.key.startswith("F") and self.key[1:].isdigit():
            f_key_name = f"WXK_{self.key}"
            value = getattr(wx_module, f_key_name, None)
            if value is None:
                raise ValueError(f"{self.key} is not available in this wx build.")
            return value
        if self.key in _SPECIAL_KEY_CODE_NAMES:
            attr_name, fallback = _SPECIAL_KEY_CODE_NAMES[self.key]
            return getattr(wx_module, attr_name, fallback)
        raise ValueError(f"Unsupported shortcut key {self.key!r}.")


WINDOW_TRAY_SHORTCUTS: tuple[ShortcutPreference, ...] = (
    ShortcutPreference(
        setting_name="shortcut_show_main_window",
        label="Show main window shortcut",
        default=DEFAULT_SHOW_MAIN_WINDOW_SHORTCUT,
        handler_name="_on_show_main_window_shortcut",
    ),
    ShortcutPreference(
        setting_name="shortcut_hide_main_window",
        label="Hide window shortcut",
        default=DEFAULT_HIDE_MAIN_WINDOW_SHORTCUT,
        handler_name="_on_hide_main_window_shortcut",
    ),
    ShortcutPreference(
        setting_name="shortcut_read_tray_info",
        label="Read tray information shortcut",
        default=DEFAULT_READ_TRAY_INFO_SHORTCUT,
        handler_name="_on_read_tray_info_shortcut",
    ),
)

WINDOW_TRAY_SHORTCUT_DEFAULTS = {
    preference.setting_name: preference.default for preference in WINDOW_TRAY_SHORTCUTS
}

RESERVED_SHORTCUTS = {
    "Ctrl+R": "refresh the weather",
    "Ctrl+L": "add a location",
    "Ctrl+D": "remove a location",
    "Ctrl+H": "open Weather History",
    "Ctrl+1": "focus Current Conditions",
    "Ctrl+2": "focus Hourly / Near-Term",
    "Ctrl+3": "focus Daily Forecast",
    "Ctrl+4": "focus Alerts",
    "Ctrl+5": "focus Event Center",
    "Ctrl+S": "open Settings",
    "Ctrl+Q": "exit AccessiWeather",
    "Ctrl+E": "open Explain Conditions",
    "Ctrl+T": "open Weather Assistant",
    "Ctrl+Shift+R": "open NOAA Weather Radio",
    "Escape": "hide the window when minimize-to-tray is enabled",
    "F5": "refresh the weather",
    "F6": "cycle through top-level sections",
}


def normalize_shortcut_text(value: str | None, *, allow_empty: bool = True) -> str:
    """Return a canonical shortcut string, or an empty string when disabled."""
    text = str(value or "").strip()
    if not text:
        if allow_empty:
            return ""
        raise ValueError("Shortcut cannot be blank.")

    parts = [part.strip() for part in text.split("+")]
    if not all(parts):
        raise ValueError("Use '+' only between modifiers and the key, for example Ctrl+Shift+W.")

    raw_key = parts[-1]
    modifier_parts = parts[:-1]
    modifiers: list[str] = []
    seen_modifiers: set[str] = set()
    for part in modifier_parts:
        normalized_modifier = _MODIFIER_ALIASES.get(part.upper())
        if normalized_modifier is None:
            raise ValueError(f"Unknown modifier {part!r}. Use Ctrl, Alt, or Shift.")
        if normalized_modifier in seen_modifiers:
            raise ValueError(f"{normalized_modifier} appears more than once.")
        seen_modifiers.add(normalized_modifier)
        modifiers.append(normalized_modifier)

    normalized_key = _normalize_key_token(raw_key)
    ordered_modifiers = [name for name in _VISIBLE_MODIFIER_ORDER if name in modifiers]
    return "+".join([*ordered_modifiers, normalized_key])


def parse_shortcut_text(value: str | None) -> ShortcutBinding | None:
    """Parse a canonical or user-entered shortcut string."""
    normalized = normalize_shortcut_text(value, allow_empty=True)
    if not normalized:
        return None
    parts = normalized.split("+")
    return ShortcutBinding(modifiers=tuple(parts[:-1]), key=parts[-1])


def resolve_shortcut_binding(value: str | None, *, default: str) -> ShortcutBinding | None:
    """Parse a user preference, falling back to the default when malformed."""
    try:
        return parse_shortcut_text(value)
    except ValueError:
        return parse_shortcut_text(default)


def _normalize_key_token(token: str) -> str:
    token = token.strip()
    if not token:
        raise ValueError("Shortcut needs a key after the modifiers.")

    special = _SPECIAL_KEYS.get(token.upper())
    if special is not None:
        return special

    if len(token) == 1 and token.isalnum():
        return token.upper()

    upper_token = token.upper()
    if upper_token.startswith("F") and upper_token[1:].isdigit():
        number = int(upper_token[1:])
        if 1 <= number <= 24:
            return f"F{number}"
        raise ValueError("Function-key shortcuts must be between F1 and F24.")

    raise ValueError("Shortcut keys must be letters, numbers, F1-F24, Tab, Space, or Escape.")

File: rust/crates/aw-core/src/shortcut_preferences.rs
//! Configurable window and tray shortcuts (`shortcut_preferences.py`) and the
//! canonical spelling of the NOAA Weather Radio hotkey
//! (`global_hotkeys.normalize_hotkey`).

use crate::settings::{
    DEFAULT_HIDE_MAIN_WINDOW_SHORTCUT, DEFAULT_READ_TRAY_INFO_SHORTCUT,
    DEFAULT_SHOW_MAIN_WINDOW_SHORTCUT,
};

/// `ShortcutPreference`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ShortcutPreference {
    pub setting_name: &'static str,
    pub label: &'static str,
    pub default: &'static str,
}

/// `WINDOW_TRAY_SHORTCUTS`, in validation order.
pub const WINDOW_TRAY_SHORTCUTS: [ShortcutPreference; 3] = [
    ShortcutPreference {
        setting_name: "shortcut_show_main_window",
        label: "Show main window shortcut",
        default: DEFAULT_SHOW_MAIN_WINDOW_SHORTCUT,
    },
    ShortcutPreference {
        setting_name: "shortcut_hide_main_window",
        label: "Hide window shortcut",
        default: DEFAULT_HIDE_MAIN_WINDOW_SHORTCUT,
    },
    ShortcutPreference {
        setting_name: "shortcut_read_tray_info",
        label: "Read tray information shortcut",
        default: DEFAULT_READ_TRAY_INFO_SHORTCUT,
    },
];

/// `RESERVED_SHORTCUTS`: in-app keys a window/tray shortcut may not take.
const RESERVED_SHORTCUTS: [(&str, &str); 19] = [
    ("Alt+F4", "close the window"),
    ("F2", "edit a location"),
    ("Ctrl+R", "refresh the weather"),
    ("Ctrl+L", "add a location"),
    ("Ctrl+D", "remove a location"),
    ("Ctrl+H", "open Weather History"),
    ("Ctrl+1", "focus Current Conditions"),
    ("Ctrl+2", "focus Hourly / Near-Term"),
    ("Ctrl+3", "focus Daily Forecast"),
    ("Ctrl+4", "focus Alerts"),
    ("Ctrl+5", "focus Event Center"),
    ("Ctrl+S", "open Settings"),
    ("Ctrl+Q", "exit AccessiWeather"),
    ("Ctrl+E", "open Explain Conditions"),
    ("Ctrl+T", "open Weather Assistant"),
    ("Ctrl+N", "open NOAA Weather Radio"),
    ("Escape", "hide the window when minimize-to-tray is enabled"),
    ("F5", "refresh the weather"),
    ("F6", "cycle through top-level sections"),
];

/// What a reserved shortcut already does, e.g. "open Settings".
pub fn reserved_shortcut_use(normalized: &str) -> Option<&'static str> {
    RESERVED_SHORTCUTS
        .iter()
        .find(|(combo, _)| *combo == normalized)
        .map(|(_, used_for)| *used_for)
}

/// Python's `repr()` of a string, as used in error messages.
pub fn py_repr(s: &str) -> String {
    let quote = if s.contains('\'') && !s.contains('"') {
        '"'
    } else {
        '\''
    };
    let mut out = String::with_capacity(s.len() + 2);
    out.push(quote);
    for c in s.chars() {
        match c {
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if c == quote => {
                out.push('\\');
                out.push(c);
            }
            c if (c as u32) < 0x20 || c as u32 == 0x7f => {
                out.push_str(&format!("\\x{:02x}", c as u32));
            }
            c => out.push(c),
        }
    }
    out.push(quote);
    out
}

/// `normalize_shortcut_text`: the canonical spelling ("Ctrl+Alt+Shift+W"),
/// "" for a blank field when `allow_empty`, or the user-facing problem.
pub fn normalize_shortcut_text(value: &str, allow_empty: bool) -> Result<String, String> {
    let text = value.trim();
    if text.is_empty() {
        return if allow_empty {
            Ok(String::new())
        } else {
            Err("Shortcut cannot be blank.".into())
        };
    }
    let parts: Vec<&str> = text.split('+').map(str::trim).collect();
    if parts.iter().any(|p| p.is_empty()) {
        return Err(
            "Use '+' only between modifiers and the key, for example Ctrl+Alt+Shift+W.".into(),
        );
    }
    let (raw_key, modifier_parts) = parts.split_last().expect("split yields one part");
    let mut modifiers: Vec<&str> = Vec::new();
    for part in modifier_parts {
        let normalized = match part.to_uppercase().as_str() {
            "ALT" | "OPTION" => "Alt",
            "CTRL" | "CONTROL" | "CMD" | "COMMAND" => "Ctrl",
            "SHIFT" => "Shift",
            _ => {
                return Err(format!(
                    "Unknown modifier {}. Use Ctrl, Alt, or Shift.",
                    py_repr(part)
                ))
            }
        };
        if modifiers.contains(&normalized) {
            return Err(format!("{normalized} appears more than once."));
        }
        modifiers.push(normalized);
    }
    let key = normalize_key_token(raw_key)?;
    let mut out: Vec<String> = ["Ctrl", "Alt", "Shift"]
        .into_iter()
        .filter(|m| modifiers.contains(m))
        .map(String::from)
        .collect();
    out.push(key);
    Ok(out.join("+"))
}

fn normalize_key_token(token: &str) -> Result<String, String> {
    let token = token.trim();
    if token.is_empty() {
        return Err("Shortcut needs a key after the modifiers.".into());
    }
    let upper = token.to_uppercase();
    match upper.as_str() {
        "ESC" | "ESCAPE" => return Ok("Escape".into()),
        "SPACE" => return Ok("Space".into()),
        "TAB" => return Ok("Tab".into()),
        _ => {}
    }
    let mut chars = token.chars();
    if let (Some(c), None) = (chars.next(), chars.next()) {
        if c.is_alphanumeric() {
            return Ok(upper);
        }
    }
    if let Some(digits) = function_key_digits(&upper) {
        return match digits.parse::<u32>() {
            Ok(n @ 1..=24) => Ok(format!("F{n}")),
            _ => Err("Function-key shortcuts must be between F1 and F24.".into()),
        };
    }
    Err("Shortcut keys must be letters, numbers, F1-F24, Tab, Space, or Escape.".into())
}

/// The digits after a leading "F" (`s[1:].isdigit()`).
fn function_key_digits(upper: &str) -> Option<&str> {
    let digits = upper.strip_prefix('F')?;
    (!digits.is_empty() && digits.chars().all(|c| c.is_ascii_digit())).then_some(digits)
}

/// wx.WXK_F1: global hotkey key codes above it are function keys.
const WXK_F1: u32 = 340;
const MAX_FUNCTION_KEY: u32 = 24;

/// `global_hotkeys.normalize_hotkey`: the canonical spelling of a system-wide
/// hotkey ("ctrl+alt+shift+r" -> "Ctrl+Alt+Shift+R"), or the
/// `HotkeyParseError` message. Unlike window shortcuts it needs a modifier
/// and also accepts Win.
pub fn normalize_hotkey(combo: &str) -> Result<String, String> {
    let parts: Vec<&str> = combo.split('+').map(str::trim).collect();
    if parts.len() < 2 || parts.iter().any(|p| p.is_empty()) {
        return Err(format!(
            "Hotkey must be modifiers plus a key, got {}",
            py_repr(combo)
        ));
    }
    let (key, modifier_parts) = parts.split_last().expect("at least two parts");
    // Ctrl, Alt, Shift, Win: canonical order no matter how they were typed.
    let mut flags = [false; 4];
    for part in modifier_parts {
        let index = match part.to_lowercase().as_str() {
            "ctrl" | "control" => 0,
            "alt" => 1,
            "shift" => 2,
            "win" | "super" | "cmd" | "command" => 3,
            _ => {
                return Err(format!(
                    "Unknown modifier {} in hotkey {}",
                    py_repr(part),
                    py_repr(combo)
                ))
            }
        };
        flags[index] = true;
    }
    let keycode = parse_hotkey_key(key).ok_or_else(|| {
        format!(
            "Unsupported key {} in hotkey {}",
            py_repr(key),
            py_repr(combo)
        )
    })?;
    let mut out: Vec<String> = ["Ctrl", "Alt", "Shift", "Win"]
        .into_iter()
        .zip(flags)
        .filter(|(_, on)| *on)
        .map(|(label, _)| label.to_string())
        .collect();
    // `_format_key`: codes in the function-key range print as F-keys.
    out.push(if (WXK_F1..WXK_F1 + MAX_FUNCTION_KEY).contains(&keycode) {
        format!("F{}", keycode - WXK_F1 + 1)
    } else {
        char::from_u32(keycode).unwrap_or_default().to_string()
    });
    Ok(out.join("+"))
}

fn parse_hotkey_key(key: &str) -> Option<u32> {
    let upper = key.to_uppercase();
    let mut chars = upper.chars();
    if let (Some(c), None) = (chars.next(), chars.next()) {
        if c.is_alphabetic() || c.is_ascii_digit() {
            return Some(c as u32);
        }
    }
    let n: u32 = function_key_digits(&upper)?.parse().ok()?;
    (1..=MAX_FUNCTION_KEY)
        .contains(&n)
        .then_some(WXK_F1 + n - 1)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn repr_matches_python_quoting() {
        assert_eq!(py_repr("Hyper"), "'Hyper'");
        assert_eq!(py_repr("it's"), "\"it's\"");
        assert_eq!(py_repr("a'b\"c"), "'a\\'b\"c'");
        assert_eq!(py_repr("tab\there\\"), "'tab\\there\\\\'");
    }
}
